Document Leak Occurs when opening Context Menu

NathanSolomon733 opened the pull request:
#### a6ce7b296295f0b388d6c7ea235bdc2490d13508
<pre>
Document Leak Occurs when opening Context Menu
<a href="https://bugs.webkit.org/show_bug.cgi?id=284189">https://bugs.webkit.org/show_bug.cgi?id=284189</a>
<a href="https://rdar.apple.com/137308000">rdar://137308000</a>

Reviewed by Wenson Hsieh.
When a context click occurs, the ContextMenu is created.
The Page currently owns the ContextMenuController, which owns the
ContextMenu. The context menu&apos;s context strongly holds onto the event
target of the context click, as well as HitTestResults which hold
Refs to Nodes.
The document that the context click occured
on can&apos;t be destructed because of this strong reference to the Node that
is held by the context menu&apos;s context. Because a new context object
is already being created for every new context click, ensure that we
destroy the held context when dismissing the context menu so that
the Page does not indirectly hold onto a node during document
destruction.

* LayoutTests/editing/mac/spelling/disable-automatic-spelling-correction-context-menu-item-does-not-leak-expected.txt: Added.
* LayoutTests/editing/mac/spelling/disable-automatic-spelling-correction-context-menu-item-does-not-leak.html: Added.
* LayoutTests/editing/mac/spelling/disable-automatic-spelling-correction-context-menu-item.html:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/resources/document-leak-test.js:
(async runDocumentLeakTestSynchronously):
(cleanIframeUponMessageReceived):
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::didDismissContextMenu):
* Source/WebCore/page/ContextMenuController.h:
(WebCore::ContextMenuController::setContextMenuContext): Deleted.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::resetToConsistentState):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didDismissContextMenu):

Canonical link: <a href="https://commits.webkit.org/288976@main">https://commits.webkit.org/288976@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77738ad68910b019f0d51f4679ddbf0c01a2a9b0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84127 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3745 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38428 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89202 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35135 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86212 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3836 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11714 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65425 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23267 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87173 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2861 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76440 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45720 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2803 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30666 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34184 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73750 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31426 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90583 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11392 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8266 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73882 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11616 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72270 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73090 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18241 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17393 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/2739 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11344 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16816 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11192 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14668 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12964 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

NathanSolomon733 commented on Source/WebCore/page/ContextMenuContext.h:
```diff
@@ -82,7 +81,7 @@ class ContextMenuContext {
     Type m_type { Type::ContextMenu };
     HitTestResult m_hitTestResult;
-    RefPtr<Event> m_event;
+    WeakPtr<Event> m_event;
```
I can't explain yet why this needs to be a WeakPtr, but I stop seeing leaked documents when running `notifyutil -p org.WebKit.lowMemory && notifyutil -p com.apple.WebKit.showAllDocuments` only if this is made WeakPtr.

rreno replied:
I'm guessing it's because the Event holds a RefPtr to an EventTarget. Nodes are EventTargets.

rreno replied:
You probably want to figure out the correct ownership semantics here. Making everything a weak pointer would probably solve all ref cycles but it would also be incorrect lifetime management.

NathanSolomon733 commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -1775,6 +1776,14 @@ void FrameSelection::willBeRemovedFromFrame()
+        page->contextMenuController().destroyContext();
+#endif
+    }
 }
```
This code is added just to make writing a document leak test using iframes possible for this kind of leak. The focused element keeps a strong ref to a node if a context click occurs within an input element, so we need to clear it.
We also destroy the context here because EventSendingController::contextClick() does not close the context menu unlike a real user would, so the context stays alive. Other tests also assume that the menu is open to test actions like clicking on the menu. 
It's not ideal that we are adding code here just to make this test work right, so I am open to suggestions on how to fix this. We can destroy the context in `Internals::resetToConsistentState(Page& page)`, which makes --world-leaks tool not falsely fail a test, but that doesn't fix an iframe-based test because resetToConsistentState is only called once after the test case has finished running and not for every iframe.

NathanSolomon733 replied:
My preference would be to modify` spelling/disable-automatic-spelling-correction-context-menu-item.html` to do another click somewhere to close the context menu after opening it and also modify Internals::resetToConsistentState() to destroy the context. 
That way, we do not get falsely flagged document leaks for tests that call JS contextClick() in the future and we can also do an accurate iframe-based document leak test, but just for this specific test case.

whsieh reviewed (approved):
Have we also considered resetting the context on main frame navigation? (There's similar cleanup logic already, in `WebPage::didCommitLoad`)

NathanSolomon733 committed:
Document Leak Occurs when opening Context Menu
https://bugs.webkit.org/show_bug.cgi?id=284189
rdar://137308000

Reviewed by Wenson Hsieh.
When a context click occurs, the ContextMenu is created.
The Page currently owns the ContextMenuController, which owns the
ContextMenu. The context menu's context strongly holds onto the event
target of the context click, as well as HitTestResults which hold
Refs to Nodes.
The document that the context click occured
on can't be destructed because of this strong reference to the Node that
is held by the context menu's context. Because a new context object
is already being created for every new context click, ensure that we
destroy the held context when dismissing the context menu so that
the Page does not indirectly hold onto a node during document
destruction.

* LayoutTests/editing/mac/spelling/disable-automatic-spelling-correction-context-menu-item-does-not-leak-expected.txt: Added.
* LayoutTests/editing/mac/spelling/disable-automatic-spelling-correction-context-menu-item-does-not-leak.html: Added.
* LayoutTests/editing/mac/spelling/disable-automatic-spelling-correction-context-menu-item.html:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/resources/document-leak-test.js:
(async runDocumentLeakTestSynchronously):
(cleanIframeUponMessageReceived):
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::didDismissContextMenu):
* Source/WebCore/page/ContextMenuController.h:
(WebCore::ContextMenuController::setContextMenuContext): Deleted.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::resetToConsistentState):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didDismissContextMenu):

Canonical link: https://commits.webkit.org/288976@main
